**For a Cricket fan nothing is more interesting than World Cups. Even those people also watch world cup mathces who don't follow or watch normal bilateral matches. As the next big event in cricketing world in "World Cup T20" , so Let's try to make some Predictive models for Score Prediction.**



In this Project we will Predict the total score of the team based on the current situation of match based on past data for T20Is that we have. The data is taken from **cricsheet** which provides ball to ball data for most of the matches played globally.

**cricsheet** has provided all ball by ball data in **".yaml"** form, so the first task is to convert in to pandas DataFrame so that we can do further work with the data.

# Data Extraction

In [1]:
#importing the libraries
import os
import numpy as np
import pandas as pd
from yaml import safe_load              #for yaml to pandas
from tqdm import tqdm

In [2]:
# with the help of os module let's extract all filenames and store them in a list
filenames=[]
for file in os.listdir("t20s"):
    filenames.append(os.path.join("t20s",file))

In [3]:
#let's check whether filenames contains all names of files with their directory
filenames[:10]

['t20s\\1001349.yaml',
 't20s\\1001351.yaml',
 't20s\\1001353.yaml',
 't20s\\1004729.yaml',
 't20s\\1007655.yaml',
 't20s\\1007657.yaml',
 't20s\\1007659.yaml',
 't20s\\1019979.yaml',
 't20s\\1019981.yaml',
 't20s\\1019983.yaml']

Now we will take each file individually and then after applying json normalizer to it then we will convert it to pandas dataframe with safeload. And we are adding one more column with match id like match no. 1,2 etc.

In [4]:
dataframe = pd.DataFrame()
count=1
for file in tqdm(filenames):
    with open(file,'r') as data:
        df = pd.json_normalize(safe_load(data))
        df['match_id']=count
        dataframe=dataframe.append(df)
        count+=1

return dataframe

100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉| 1432/1433 [11:05<00:00,  2.15it/s]


NotImplementedError: 

In [5]:
dataframe

innings  meta.data_version  \
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...                0.9   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...                0.9   
..                                                ...                ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...                0.9   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...                0.9   
0   [{'1st innings': {'team': 'Netherlands', 'deli...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...                0.9   

   meta.created  meta.revision    info.dates info.gender info.match_type  \
0    2017-02-18              2  [2017-02-17]        male             T20   
0    2017-02-19              2  [2017-02-19]        male             T20   
0    2017-02-23              1  [2017-02-22]        male             T20   
0    2016-09-12              1  [2016-09-05]        male             T20   
0    2016-06-19              1  [2016-06-18]        male             T20   
..          ...            ...           ...         ...             ...   
0    2016-03-05              2  [2016-03-04]        male             T20   
0    2016-03-08              1  [2016-03-06]        male             T20   
0    2016-02-03              1  [2016-02-03]        male             T20   
0    2016-09-12              1  [2016-09-06]        male             T20   
0    2016-09-12              1  [2016-09-09]        male             T20   

    info.outcome.by.wickets info.outcome.winner  info.overs  ...  \
0                       5.0           Sri Lanka          20  ...   
0                       2.0           Sri Lanka          20  ...   
0                       NaN           Australia          20  ...   
0                       NaN           Hong Kong          20  ...   
0                       NaN            Zimbabwe          20  ...   
..                      ...                 ...         ...  ...   
0                       6.0            Pakistan          20  ...   
0                       8.0               India          20  ...   
0                       NaN         Netherlands          20  ...   
0                       NaN           Australia          20  ...   
0                       4.0           Australia          20  ...   

   info.outcome.by.runs info.match_type_number info.neutral_venue  \
0                   NaN                    NaN                NaN   
0                   NaN                    NaN                NaN   
0                  41.0                    NaN                NaN   
0                  40.0                    NaN                NaN   
0                   2.0                    NaN                NaN   
..                  ...                    ...                ...   
0                   NaN                    NaN                1.0   
0                   NaN                    NaN                NaN   
0                  84.0                    NaN                NaN   
0                  85.0                    NaN                NaN   
0                   NaN                    NaN                NaN   

   info.outcome.method info.outcome.result info.outcome.eliminator  \
0                  NaN                 NaN                     NaN   
0                  NaN                 NaN                     NaN   
0                  NaN                 NaN                     NaN   
0                  NaN                 NaN                     NaN   
0                  NaN                 NaN                     NaN   
..                 ...                 ...                     ...   
0                  NaN                 NaN                     NaN   
0             

As we can see there are 1432 rows  means we have this much data and this includes all t20 maches including women t20 also. **innings** column is kind of mose useful thing for us as it contains all ball by ball data.

In [6]:
#let's check what all columns we have in dataset 
dataframe.columns

Index(['innings', 'meta.data_version', 'meta.created', 'meta.revision',
       'info.dates', 'info.gender', 'info.match_type',
       'info.outcome.by.wickets', 'info.outcome.winner', 'info.overs',
       'info.player_of_match', 'info.teams', 'info.toss.decision',
       'info.toss.winner', 'info.umpires', 'info.venue', 'match_id',
       'info.city', 'info.outcome.by.runs', 'info.match_type_number',
       'info.neutral_venue', 'info.outcome.method', 'info.outcome.result',
       'info.outcome.eliminator', 'info.supersubs.New Zealand',
       'info.supersubs.South Africa', 'info.bowl_out',
       'info.outcome.bowl_out'],
      dtype='object')

We can see there are many columns or features are there which might not be that much of importance and they can make our predictive model more complex without adding much value to it. So we will drop some features.

In [7]:
#first let's make a backup
backupt20s = dataframe.copy()

In [8]:
dataframe.drop(columns=['meta.data_version','meta.created','meta.revision','info.outcome.bowl_out',
                       'info.bowl_out','info.supersubs.South Africa','info.supersubs.New Zealand',
                       'info.outcome.eliminator','info.outcome.method','info.neutral_venue',
                       'info.match_type_number','info.outcome.by.runs','info.outcome.by.wickets'
                        ],inplace=True)

In [9]:
dataframe

innings    info.dates  \
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
..                                                ...           ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
0   [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

   info.gender info.match_type info.outcome.winner  info.overs  \
0         male             T20           Sri Lanka          20   
0         male             T20           Sri Lanka          20   
0         male             T20           Australia          20   
0         male             T20           Hong Kong          20   
0         male             T20            Zimbabwe          20   
..         ...             ...                 ...         ...   
0         male             T20            Pakistan          20   
0         male             T20               India          20   
0         male             T20         Netherlands          20   
0         male             T20           Australia          20   
0         male             T20           Australia          20   

   info.player_of_match                           info.teams  \
0       [DAS Gunaratne]               [Australia, Sri Lanka]   
0       [DAS Gunaratne]               [Australia, Sri Lanka]   
0             [A Zampa]               [Australia, Sri Lanka]   
0                   NaN                 [Ireland, Hong Kong]   
0        [E Chigumbura]                    [Zimbabwe, India]   
..                  ...                                  ...   
0          [Umar Akmal]                [Pakistan, Sri Lanka]   
0            [S Dhawan]                  [Bangladesh, India]   
0    [Mudassar Bukhari]  [United Arab Emirates, Netherlands]   
0          [GJ Maxwell]               [Sri Lanka, Australia]   
0          [GJ Maxwell]               [Sri Lanka, Australia]   

   info.toss.decision      info.toss.winner  \
0               field             Sri Lanka   
0               field             Sri Lanka   
0               field             Sri Lanka   
0                 bat             Hong Kong   
0               field                 India   
..                ...                   ...   
0               field              Pakistan   
0               field                 India   
0               field  United Arab Emirates   
0               field             Sri Lanka   
0                 bat             Sri Lanka   

                          info.umpires  \
0               [MD Martell, P Wilson]   
0                [SD Fry, SJ Nogajski]   
0               [MD Martell, P Wilson]   
0                  [R Black, AJ Neill]   
0             [TJ Matibiri, RB Tiffin]   
..                                 ...   
0         [AK Chaudhary, Enamul Haque]   
0     [RSA Palliyaguruge, Shozab Raza]   
0           [CK Nandan, Sarika Prasad]   
0       [REJ Martinesz, RR Wimalasiri]   
0   [REJ Martinesz, RSA Palliyaguruge]   

                                 info.venue  match_id    info.city  \
0                  Melbourne Cricket Ground         1          NaN   
0            Simonds Stadium, South Geelong         2     Victoria   
0                             Adelaide Oval         3          NaN   
0         Bready Cricket Club, Magheramason         4  Londonderry   
0                        Harare Sports Club         5          NaN   
..                                      ...       ...          ...   
0             Shere Bangla National Stadium      1428       Mirpur   
0

In [10]:
dataframe['info.gender'].value_counts()

male      966
female    466
Name: info.gender, dtype: int64

So we have data for both men and women out of which 966 matche data is for men T20s. in this project we are intrested in Men's T20 score so we will consider only that data.

In [11]:
dataframe=dataframe[dataframe['info.gender']=='male']
dataframe.drop(columns=['info.gender'],inplace=True)
dataframe

C:\Users\Mahesh\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


innings    info.dates  \
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
..                                                ...           ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
0   [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

   info.match_type info.outcome.winner  info.overs info.player_of_match  \
0              T20           Sri Lanka          20      [DAS Gunaratne]   
0              T20           Sri Lanka          20      [DAS Gunaratne]   
0              T20           Australia          20            [A Zampa]   
0              T20           Hong Kong          20                  NaN   
0              T20            Zimbabwe          20       [E Chigumbura]   
..             ...                 ...         ...                  ...   
0              T20            Pakistan          20         [Umar Akmal]   
0              T20               India          20           [S Dhawan]   
0              T20         Netherlands          20   [Mudassar Bukhari]   
0              T20           Australia          20         [GJ Maxwell]   
0              T20           Australia          20         [GJ Maxwell]   

                             info.teams info.toss.decision  \
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                  [Ireland, Hong Kong]                bat   
0                     [Zimbabwe, India]              field   
..                                  ...                ...   
0                 [Pakistan, Sri Lanka]              field   
0                   [Bangladesh, India]              field   
0   [United Arab Emirates, Netherlands]              field   
0                [Sri Lanka, Australia]              field   
0                [Sri Lanka, Australia]                bat   

        info.toss.winner                        info.umpires  \
0              Sri Lanka              [MD Martell, P Wilson]   
0              Sri Lanka               [SD Fry, SJ Nogajski]   
0              Sri Lanka              [MD Martell, P Wilson]   
0              Hong Kong                 [R Black, AJ Neill]   
0                  India            [TJ Matibiri, RB Tiffin]   
..                   ...                                 ...   
0               Pakistan        [AK Chaudhary, Enamul Haque]   
0                  India    [RSA Palliyaguruge, Shozab Raza]   
0   United Arab Emirates          [CK Nandan, Sarika Prasad]   
0              Sri Lanka      [REJ Martinesz, RR Wimalasiri]   
0              Sri Lanka  [REJ Martinesz, RSA Palliyaguruge]   

                                 info.venue  match_id    info.city  \
0                  Melbourne Cricket Ground         1          NaN   
0            Simonds Stadium, South Geelong         2     Victoria   
0                             Adelaide Oval         3          NaN   
0         Bready Cricket Club, Magheramason         4  Londonderry   
0                        Harare Sports Club         5          NaN   
..                                      ...       ...          ...   
0             Shere Bangla National Stadium      1428       Mirpur   
0             Shere Bangla National Stadium      1429       Mirpur   
0                               ICC Academy      1430        Dubai   
0   Pallekele International Cricket Stadium      1431          NaN   
0       

In [12]:
dataframe['info.match_type'].value_counts()

T20    966
Name: info.match_type, dtype: int64

In [13]:
dataframe['info.overs'].value_counts()

20    963
50      3
Name: info.overs, dtype: int64

As we can see that match_type  contains all same value and overs contains only three data where overs are 50, as we are intrested in T20s so we will drop those 50 overs rows.

In [14]:
dataframe=dataframe[dataframe['info.overs']==20]
dataframe.drop(columns=['info.match_type','info.overs'],inplace=True)
dataframe

C:\Users\Mahesh\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


innings    info.dates  \
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
..                                                ...           ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
0   [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

   info.outcome.winner info.player_of_match  \
0            Sri Lanka      [DAS Gunaratne]   
0            Sri Lanka      [DAS Gunaratne]   
0            Australia            [A Zampa]   
0            Hong Kong                  NaN   
0             Zimbabwe       [E Chigumbura]   
..                 ...                  ...   
0             Pakistan         [Umar Akmal]   
0                India           [S Dhawan]   
0          Netherlands   [Mudassar Bukhari]   
0            Australia         [GJ Maxwell]   
0            Australia         [GJ Maxwell]   

                             info.teams info.toss.decision  \
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                  [Ireland, Hong Kong]                bat   
0                     [Zimbabwe, India]              field   
..                                  ...                ...   
0                 [Pakistan, Sri Lanka]              field   
0                   [Bangladesh, India]              field   
0   [United Arab Emirates, Netherlands]              field   
0                [Sri Lanka, Australia]              field   
0                [Sri Lanka, Australia]                bat   

        info.toss.winner                        info.umpires  \
0              Sri Lanka              [MD Martell, P Wilson]   
0              Sri Lanka               [SD Fry, SJ Nogajski]   
0              Sri Lanka              [MD Martell, P Wilson]   
0              Hong Kong                 [R Black, AJ Neill]   
0                  India            [TJ Matibiri, RB Tiffin]   
..                   ...                                 ...   
0               Pakistan        [AK Chaudhary, Enamul Haque]   
0                  India    [RSA Palliyaguruge, Shozab Raza]   
0   United Arab Emirates          [CK Nandan, Sarika Prasad]   
0              Sri Lanka      [REJ Martinesz, RR Wimalasiri]   
0              Sri Lanka  [REJ Martinesz, RSA Palliyaguruge]   

                                 info.venue  match_id    info.city  \
0                  Melbourne Cricket Ground         1          NaN   
0            Simonds Stadium, South Geelong         2     Victoria   
0                             Adelaide Oval         3          NaN   
0         Bready Cricket Club, Magheramason         4  Londonderry   
0                        Harare Sports Club         5          NaN   
..                                      ...       ...          ...   
0             Shere Bangla National Stadium      1428       Mirpur   
0             Shere Bangla National Stadium      1429       Mirpur   
0                               ICC Academy      1430        Dubai   
0   Pallekele International Cricket Stadium      1431          NaN   
0                       R Premadasa Stadium      1432      Colombo   

   info.outcome.result  
0                  NaN  
0                  NaN  
0                  NaN  
0                  NaN  
0                  NaN  
..                 ...  
0                  NaN  
0                  NaN  
0                  NaN  
0                  NaN 

In [15]:
#As data extraction is taking much time so we don't want to lose our progress. We will pickle out whatever we have done
import pickle
pickle.dump(dataframe,open('dataset1.pkl','wb'))

In [16]:
matches = pickle.load(open('dataset1.pkl','rb'))
matches

innings    info.dates  \
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
..                                                ...           ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
0   [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

   info.outcome.winner info.player_of_match  \
0            Sri Lanka      [DAS Gunaratne]   
0            Sri Lanka      [DAS Gunaratne]   
0            Australia            [A Zampa]   
0            Hong Kong                  NaN   
0             Zimbabwe       [E Chigumbura]   
..                 ...                  ...   
0             Pakistan         [Umar Akmal]   
0                India           [S Dhawan]   
0          Netherlands   [Mudassar Bukhari]   
0            Australia         [GJ Maxwell]   
0            Australia         [GJ Maxwell]   

                             info.teams info.toss.decision  \
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                  [Ireland, Hong Kong]                bat   
0                     [Zimbabwe, India]              field   
..                                  ...                ...   
0                 [Pakistan, Sri Lanka]              field   
0                   [Bangladesh, India]              field   
0   [United Arab Emirates, Netherlands]              field   
0                [Sri Lanka, Australia]              field   
0                [Sri Lanka, Australia]                bat   

        info.toss.winner                        info.umpires  \
0              Sri Lanka              [MD Martell, P Wilson]   
0              Sri Lanka               [SD Fry, SJ Nogajski]   
0              Sri Lanka              [MD Martell, P Wilson]   
0              Hong Kong                 [R Black, AJ Neill]   
0                  India            [TJ Matibiri, RB Tiffin]   
..                   ...                                 ...   
0               Pakistan        [AK Chaudhary, Enamul Haque]   
0                  India    [RSA Palliyaguruge, Shozab Raza]   
0   United Arab Emirates          [CK Nandan, Sarika Prasad]   
0              Sri Lanka      [REJ Martinesz, RR Wimalasiri]   
0              Sri Lanka  [REJ Martinesz, RSA Palliyaguruge]   

                                 info.venue  match_id    info.city  \
0                  Melbourne Cricket Ground         1          NaN   
0            Simonds Stadium, South Geelong         2     Victoria   
0                             Adelaide Oval         3          NaN   
0         Bready Cricket Club, Magheramason         4  Londonderry   
0                        Harare Sports Club         5          NaN   
..                                      ...       ...          ...   
0             Shere Bangla National Stadium      1428       Mirpur   
0             Shere Bangla National Stadium      1429       Mirpur   
0                               ICC Academy      1430        Dubai   
0   Pallekele International Cricket Stadium      1431          NaN   
0                       R Premadasa Stadium      1432      Colombo   

   info.outcome.result  
0                  NaN  
0                  NaN  
0                  NaN  
0                  NaN  
0                  NaN  
..                 ...  
0                  NaN  
0                  NaN  
0                  NaN  
0                  NaN 

The most important and trickiest task in this project or data analysis is to extract ball by ball data which is in **innings** column

In [17]:
#Let's have a look at innings
matches.iloc[0]['innings']

[{'1st innings': {'team': 'Australia',
   'deliveries': [{0.1: {'batsman': 'AJ Finch',
      'bowler': 'SL Malinga',
      'non_striker': 'M Klinger',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.2: {'batsman': 'AJ Finch',
      'bowler': 'SL Malinga',
      'non_striker': 'M Klinger',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.3: {'batsman': 'AJ Finch',
      'bowler': 'SL Malinga',
      'non_striker': 'M Klinger',
      'runs': {'batsman': 1, 'extras': 0, 'total': 1}}},
    {0.4: {'batsman': 'M Klinger',
      'bowler': 'SL Malinga',
      'non_striker': 'AJ Finch',
      'runs': {'batsman': 2, 'extras': 0, 'total': 2}}},
    {0.5: {'batsman': 'M Klinger',
      'bowler': 'SL Malinga',
      'non_striker': 'AJ Finch',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.6: {'batsman': 'M Klinger',
      'bowler': 'SL Malinga',
      'non_striker': 'AJ Finch',
      'runs': {'batsman': 3, 'extras': 0, 'total': 3}}},
    {1.1: {'batsm

So we see that one value at innings column contain two dictionaries, one for first inning and another for second innings.We are making project for innings score so either we can take first innings or second innings. Here I am using first inning data

In [18]:
#let's have a look at first inning for a particular match
matches.iloc[0]['innings'][0]['1st innings']

{'team': 'Australia',
 'deliveries': [{0.1: {'batsman': 'AJ Finch',
    'bowler': 'SL Malinga',
    'non_striker': 'M Klinger',
    'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
  {0.2: {'batsman': 'AJ Finch',
    'bowler': 'SL Malinga',
    'non_striker': 'M Klinger',
    'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
  {0.3: {'batsman': 'AJ Finch',
    'bowler': 'SL Malinga',
    'non_striker': 'M Klinger',
    'runs': {'batsman': 1, 'extras': 0, 'total': 1}}},
  {0.4: {'batsman': 'M Klinger',
    'bowler': 'SL Malinga',
    'non_striker': 'AJ Finch',
    'runs': {'batsman': 2, 'extras': 0, 'total': 2}}},
  {0.5: {'batsman': 'M Klinger',
    'bowler': 'SL Malinga',
    'non_striker': 'AJ Finch',
    'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
  {0.6: {'batsman': 'M Klinger',
    'bowler': 'SL Malinga',
    'non_striker': 'AJ Finch',
    'runs': {'batsman': 3, 'extras': 0, 'total': 3}}},
  {1.1: {'batsman': 'M Klinger',
    'bowler': 'KMDN Kulasekara',
    'non_striker

1st innings itself is dictionary containing teams and deliveries as keys. Team is quite simple, let's have a look at deliveries

In [19]:
matches.iloc[0]['innings'][0]['1st innings']['deliveries']

[{0.1: {'batsman': 'AJ Finch',
   'bowler': 'SL Malinga',
   'non_striker': 'M Klinger',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.2: {'batsman': 'AJ Finch',
   'bowler': 'SL Malinga',
   'non_striker': 'M Klinger',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.3: {'batsman': 'AJ Finch',
   'bowler': 'SL Malinga',
   'non_striker': 'M Klinger',
   'runs': {'batsman': 1, 'extras': 0, 'total': 1}}},
 {0.4: {'batsman': 'M Klinger',
   'bowler': 'SL Malinga',
   'non_striker': 'AJ Finch',
   'runs': {'batsman': 2, 'extras': 0, 'total': 2}}},
 {0.5: {'batsman': 'M Klinger',
   'bowler': 'SL Malinga',
   'non_striker': 'AJ Finch',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.6: {'batsman': 'M Klinger',
   'bowler': 'SL Malinga',
   'non_striker': 'AJ Finch',
   'runs': {'batsman': 3, 'extras': 0, 'total': 3}}},
 {1.1: {'batsman': 'M Klinger',
   'bowler': 'KMDN Kulasekara',
   'non_striker': 'AJ Finch',
   'runs': {'batsman': 0, 'extras': 0, 'total': 

So delivers contains a list where the list contains dictionary with key value pairs of **bowl**, ""**batsman**,**bowler**,**non-striker**,**runs**"".

# Now we have to extract ball by ball data from it


In [20]:
matches.isnull().sum()

innings                   0
info.dates                0
info.outcome.winner      36
info.player_of_match     98
info.teams                0
info.toss.decision        0
info.toss.winner          0
info.umpires             10
info.venue                0
match_id                  0
info.city               138
info.outcome.result     927
dtype: int64

In [21]:
# Code to extact ball by ball data from innings column and add it to dataframe 
count=0
df_delivery=pd.DataFrame()
for index,row in matches.iterrows():
    #if count in [75,108,150,180,268,360,443,458,584,748,982,1052,1111,1226,1345]:
       # count+=1
       # continue
    count+=1
    ball_of_match = []
    batsman = []
    bowler = []
    runs = []
    player_dismissed = []
    teams = []
    batting_team = []
    match_id = []
    city = []
    venue = []
    for ball in row['innings'][0]['1st innings']['deliveries']:
        for key in ball.keys():
            match_id.append(count)
            batting_team.append(row['innings'][0]['1st innings']['team'])
            teams.append(row['info.teams'])
            ball_of_match.append(key)
            batsman.append(ball[key]['batsman'])
            bowler.append(ball[key]['bowler'])
            runs.append(ball[key]['runs']['total'])
            city.append(row['info.city'])
            venue.append(row['info.venue'])
            try:
                player_dismissed.append(ball[key]['wicket']['player_out'])
            except:
                player_dismissed.append('0')
                
    df = pd.DataFrame({
            'match_id':match_id,
            'teams':teams,
            'batting_team':batting_team,
            'ball':ball_of_match,
            'batsman':batsman,
            'bowler':bowler,
            'runs':runs,
            'player_dismissed':player_dismissed,
            'city':city,
            'venue':venue
        })
    df_delivery = df_delivery.append(df)


In [22]:
df_delivery

match_id                   teams batting_team  ball          batsman  \
0           1  [Australia, Sri Lanka]    Australia   0.1         AJ Finch   
1           1  [Australia, Sri Lanka]    Australia   0.2         AJ Finch   
2           1  [Australia, Sri Lanka]    Australia   0.3         AJ Finch   
3           1  [Australia, Sri Lanka]    Australia   0.4        M Klinger   
4           1  [Australia, Sri Lanka]    Australia   0.5        M Klinger   
..        ...                     ...          ...   ...              ...   
121       963  [Sri Lanka, Australia]    Sri Lanka  19.3  SMSM Senanayake   
122       963  [Sri Lanka, Australia]    Sri Lanka  19.4      DM de Silva   
123       963  [Sri Lanka, Australia]    Sri Lanka  19.5      DM de Silva   
124       963  [Sri Lanka, Australia]    Sri Lanka  19.6  SMSM Senanayake   
125       963  [Sri Lanka, Australia]    Sri Lanka  19.7  SMSM Senanayake   

         bowler  runs player_dismissed     city                     venue  
0    SL Malinga     0                0      NaN  Melbourne Cricket Ground  
1    SL Malinga     0                0      NaN  Melbourne Cricket Ground  
2    SL Malinga     1                0      NaN  Melbourne Cricket Ground  
3    SL Malinga     2                0      NaN  Melbourne Cricket Ground  
4    SL Malinga     0                0      NaN  Melbourne Cricket Ground  
..          ...   ...              ...      ...                       ...  
121    MA Starc     1                0  Colombo       R Premadasa Stadium  
122    MA Starc     0                0  Colombo       R Premadasa Stadium  
123    MA Starc     0      DM de Silva  Colombo       R Premadasa Stadium  
124    MA Starc     2                0  Colombo       R Premadasa Stadium  
125    MA Starc     1                0  Colombo       R Premadasa Stadium  

[116547 rows x 10 columns]

So now we have data for ball by by for the first innings and we can see that we have batting team in it, so we can make another column with bowling team and for that we have to extract the bowling team from the dataset

In [23]:
# function to get bowling team column
def bowl(row):
    for team in row['teams']:
        if team != row['batting_team']:
            return team

In [24]:
df_delivery['bowling_team'] = df_delivery.apply(bowl,axis=1)

In [25]:
df_delivery

match_id                   teams batting_team  ball          batsman  \
0           1  [Australia, Sri Lanka]    Australia   0.1         AJ Finch   
1           1  [Australia, Sri Lanka]    Australia   0.2         AJ Finch   
2           1  [Australia, Sri Lanka]    Australia   0.3         AJ Finch   
3           1  [Australia, Sri Lanka]    Australia   0.4        M Klinger   
4           1  [Australia, Sri Lanka]    Australia   0.5        M Klinger   
..        ...                     ...          ...   ...              ...   
121       963  [Sri Lanka, Australia]    Sri Lanka  19.3  SMSM Senanayake   
122       963  [Sri Lanka, Australia]    Sri Lanka  19.4      DM de Silva   
123       963  [Sri Lanka, Australia]    Sri Lanka  19.5      DM de Silva   
124       963  [Sri Lanka, Australia]    Sri Lanka  19.6  SMSM Senanayake   
125       963  [Sri Lanka, Australia]    Sri Lanka  19.7  SMSM Senanayake   

         bowler  runs player_dismissed     city                     venue  \
0    SL Malinga     0                0      NaN  Melbourne Cricket Ground   
1    SL Malinga     0                0      NaN  Melbourne Cricket Ground   
2    SL Malinga     1                0      NaN  Melbourne Cricket Ground   
3    SL Malinga     2                0      NaN  Melbourne Cricket Ground   
4    SL Malinga     0                0      NaN  Melbourne Cricket Ground   
..          ...   ...              ...      ...                       ...   
121    MA Starc     1                0  Colombo       R Premadasa Stadium   
122    MA Starc     0                0  Colombo       R Premadasa Stadium   
123    MA Starc     0      DM de Silva  Colombo       R Premadasa Stadium   
124    MA Starc     2                0  Colombo       R Premadasa Stadium   
125    MA Starc     1                0  Colombo       R Premadasa Stadium   

    bowling_team  
0      Sri Lanka  
1      Sri Lanka  
2      Sri Lanka  
3      Sri Lanka  
4      Sri Lanka  
..           ...  
121    Australia  
122    Australia  
123    Australia  
124    Australia  
125    Australia  

[116547 rows x 11 columns]

Now as we have both bowling and batting teams we can drop the teams columns

In [26]:
df_delivery.drop(columns=['teams'],inplace=True)

We should look on the data that which all teams match data it contains because we know that in major ICC events only top 10 or 8 teams will play

In [27]:
df_delivery['batting_team'].unique()


array(['Australia', 'Hong Kong', 'Zimbabwe', 'India', 'Bangladesh',
       'New Zealand', 'South Africa', 'England', 'West Indies', 'Ireland',
       'Afghanistan', 'Pakistan', 'United Arab Emirates', 'Scotland',
       'Oman', 'Papua New Guinea', 'Sri Lanka', 'Netherlands', 'Nepal',
       'Vanuatu', 'Philippines', 'United States of America', 'Germany',
       'Ghana', 'Uganda', 'Kenya', 'Namibia', 'Nigeria', 'Botswana',
       'Guernsey', 'Denmark', 'Jersey', 'Italy', 'Norway', 'Thailand',
       'Malaysia', 'Maldives', 'Singapore', 'Kuwait', 'Bermuda', 'Canada',
       'Cayman Islands', 'Portugal', 'Gibraltar', 'Spain', 'Bhutan',
       'Qatar', 'Iran', 'Belgium', 'Isle of Man', 'Bulgaria', 'Romania'],
      dtype=object)

In [28]:
teams = ['Australia','India','Bangladesh','New Zealand','South Africa','England','West Indies',
         'Afghanistan','Pakistan','Sri Lanka']

As only top 8 or 10 teams participate in major ICC events so we will consider them only

In [29]:
df_delivery = df_delivery[df_delivery['batting_team'].isin(teams)]
df_delivery = df_delivery[df_delivery['bowling_team'].isin(teams)]

In [30]:
df_delivery

match_id batting_team  ball          batsman      bowler  runs  \
0           1    Australia   0.1         AJ Finch  SL Malinga     0   
1           1    Australia   0.2         AJ Finch  SL Malinga     0   
2           1    Australia   0.3         AJ Finch  SL Malinga     1   
3           1    Australia   0.4        M Klinger  SL Malinga     2   
4           1    Australia   0.5        M Klinger  SL Malinga     0   
..        ...          ...   ...              ...         ...   ...   
121       963    Sri Lanka  19.3  SMSM Senanayake    MA Starc     1   
122       963    Sri Lanka  19.4      DM de Silva    MA Starc     0   
123       963    Sri Lanka  19.5      DM de Silva    MA Starc     0   
124       963    Sri Lanka  19.6  SMSM Senanayake    MA Starc     2   
125       963    Sri Lanka  19.7  SMSM Senanayake    MA Starc     1   

    player_dismissed     city                     venue bowling_team  
0                  0      NaN  Melbourne Cricket Ground    Sri Lanka  
1                  0      NaN  Melbourne Cricket Ground    Sri Lanka  
2                  0      NaN  Melbourne Cricket Ground    Sri Lanka  
3                  0      NaN  Melbourne Cricket Ground    Sri Lanka  
4                  0      NaN  Melbourne Cricket Ground    Sri Lanka  
..               ...      ...                       ...          ...  
121                0  Colombo       R Premadasa Stadium    Australia  
122                0  Colombo       R Premadasa Stadium    Australia  
123      DM de Silva  Colombo       R Premadasa Stadium    Australia  
124                0  Colombo       R Premadasa Stadium    Australia  
125                0  Colombo       R Premadasa Stadium    Australia  

[64741 rows x 10 columns]

In [31]:
# As we are predicting final score so we are just considering which team is permorming how in the remaining overs of innings
output = df_delivery[['match_id','batting_team','bowling_team','ball','runs','player_dismissed','city','venue']]
output

match_id batting_team bowling_team  ball  runs player_dismissed     city  \
0           1    Australia    Sri Lanka   0.1     0                0      NaN   
1           1    Australia    Sri Lanka   0.2     0                0      NaN   
2           1    Australia    Sri Lanka   0.3     1                0      NaN   
3           1    Australia    Sri Lanka   0.4     2                0      NaN   
4           1    Australia    Sri Lanka   0.5     0                0      NaN   
..        ...          ...          ...   ...   ...              ...      ...   
121       963    Sri Lanka    Australia  19.3     1                0  Colombo   
122       963    Sri Lanka    Australia  19.4     0                0  Colombo   
123       963    Sri Lanka    Australia  19.5     0      DM de Silva  Colombo   
124       963    Sri Lanka    Australia  19.6     2                0  Colombo   
125       963    Sri Lanka    Australia  19.7     1                0  Colombo   

                        venue  
0    Melbourne Cricket Ground  
1    Melbourne Cricket Ground  
2    Melbourne Cricket Ground  
3    Melbourne Cricket Ground  
4    Melbourne Cricket Ground  
..                        ...  
121       R Premadasa Stadium  
122       R Premadasa Stadium  
123       R Premadasa Stadium  
124       R Premadasa Stadium  
125       R Premadasa Stadium  

[64741 rows x 8 columns]

In [32]:
pickle.dump(output,open('dataset2.pkl','wb'))

So we have successfully extracted Useful data from the dataset. Now the next step is Feature Extraction.

# Feature Extraction

Before deep diving into feature extraction, it is very important to know what features we want in our dataset for traing model and predicting outcome.
1. batting team
2. bowling team
3. current score
4. wicket left
5. current run rate
6. city
7. score in last 5 overs
8. balls left

In [33]:
output.isnull().sum()

match_id               0
batting_team           0
bowling_team           0
ball                   0
runs                   0
player_dismissed       0
city                8671
venue                  0
dtype: int64

As we are seeing that city is containing a lot of missing values but most of time venue name itself contains city names. Let's have a look

In [35]:
output[output['city'].isnull()]['venue'].value_counts()

Dubai International Cricket Stadium        3092
Pallekele International Cricket Stadium    2066
Melbourne Cricket Ground                   1453
Sydney Cricket Ground                       749
Adelaide Oval                               498
Harare Sports Club                          372
Sharjah Cricket Stadium                     249
Sylhet International Cricket Stadium        128
Carrara Oval                                 64
Name: venue, dtype: int64

So we are seeing that the rows where city is missing, the venue name itself have the city names so we can extract the city name from venue itself.

In [36]:
cities = np.where(output['city'].isnull(),output['venue'].str.split().apply(lambda x:x[0]),output['city'])

In [37]:
output['city'] = cities

C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/1481407363.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['city'] = cities


In [38]:
output

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                0   
1           1    Australia    Sri Lanka   0.2     0                0   
2           1    Australia    Sri Lanka   0.3     1                0   
3           1    Australia    Sri Lanka   0.4     2                0   
4           1    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       963    Sri Lanka    Australia  19.3     1                0   
122       963    Sri Lanka    Australia  19.4     0                0   
123       963    Sri Lanka    Australia  19.5     0      DM de Silva   
124       963    Sri Lanka    Australia  19.6     2                0   
125       963    Sri Lanka    Australia  19.7     1                0   

          city                     venue  
0    Melbourne  Melbourne Cricket Ground  
1    Melbourne  Melbourne Cricket Ground  
2    Melbourne  Melbourne Cricket Ground  
3    Melbourne  Melbourne Cricket Ground  
4    Melbourne  Melbourne Cricket Ground  
..         ...                       ...  
121    Colombo       R Premadasa Stadium  
122    Colombo       R Premadasa Stadium  
123    Colombo       R Premadasa Stadium  
124    Colombo       R Premadasa Stadium  
125    Colombo       R Premadasa Stadium  

[64741 rows x 8 columns]

In [39]:
output.isnull().sum()

match_id            0
batting_team        0
bowling_team        0
ball                0
runs                0
player_dismissed    0
city                0
venue               0
dtype: int64

In [40]:
#As we have city so now we can remove venue
output.drop(columns=['venue'],inplace=True)
output

C:\Users\Mahesh\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


match_id batting_team bowling_team  ball  runs player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                0   
1           1    Australia    Sri Lanka   0.2     0                0   
2           1    Australia    Sri Lanka   0.3     1                0   
3           1    Australia    Sri Lanka   0.4     2                0   
4           1    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       963    Sri Lanka    Australia  19.3     1                0   
122       963    Sri Lanka    Australia  19.4     0                0   
123       963    Sri Lanka    Australia  19.5     0      DM de Silva   
124       963    Sri Lanka    Australia  19.6     2                0   
125       963    Sri Lanka    Australia  19.7     1                0   

          city  
0    Melbourne  
1    Melbourne  
2    Melbourne  
3    Melbourne  
4    Melbourne  
..         ...  
121    Colombo  
122    Colombo  
123    Colombo  
124    Colombo  
125    Colombo  

[64741 rows x 7 columns]

In [41]:
output['city'].value_counts()

Colombo          4208
Mirpur           3420
Johannesburg     3331
Dubai            3092
Auckland         2655
                 ... 
Victoria          123
Potchefstroom     122
Dharamsala        122
Ahmedabad         121
Carrara            64
Name: city, Length: 87, dtype: int64

City is a categorical data and there are 87 cities listed in data so we won't consider all. We can start considering those cities where almost 5 matches ahve been played. 5 matches first innings means atleast 600 balls.

In [42]:
eligible_cities = output['city'].value_counts()[output['city'].value_counts() > 600].index.tolist()

In [43]:
output = output[output['city'].isin(eligible_cities)]

In [44]:
output

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                0   
1           1    Australia    Sri Lanka   0.2     0                0   
2           1    Australia    Sri Lanka   0.3     1                0   
3           1    Australia    Sri Lanka   0.4     2                0   
4           1    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       963    Sri Lanka    Australia  19.3     1                0   
122       963    Sri Lanka    Australia  19.4     0                0   
123       963    Sri Lanka    Australia  19.5     0      DM de Silva   
124       963    Sri Lanka    Australia  19.6     2                0   
125       963    Sri Lanka    Australia  19.7     1                0   

          city  
0    Melbourne  
1    Melbourne  
2    Melbourne  
3    Melbourne  
4    Melbourne  
..         ...  
121    Colombo  
122    Colombo  
123    Colombo  
124    Colombo  
125    Colombo  

[51109 rows x 7 columns]

No we have to calculate current score means what was the score after a certain ball

In [45]:
output['current_score'] = output.groupby('match_id').cumsum()['runs']

C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/3899481829.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['current_score'] = output.groupby('match_id').cumsum()['runs']


In [46]:
output

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                0   
1           1    Australia    Sri Lanka   0.2     0                0   
2           1    Australia    Sri Lanka   0.3     1                0   
3           1    Australia    Sri Lanka   0.4     2                0   
4           1    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       963    Sri Lanka    Australia  19.3     1                0   
122       963    Sri Lanka    Australia  19.4     0                0   
123       963    Sri Lanka    Australia  19.5     0      DM de Silva   
124       963    Sri Lanka    Australia  19.6     2                0   
125       963    Sri Lanka    Australia  19.7     1                0   

          city  current_score  
0    Melbourne              0  
1    Melbourne              0  
2    Melbourne              1  
3    Melbourne              3  
4    Melbourne              3  
..         ...            ...  
121    Colombo            125  
122    Colombo            125  
123    Colombo            125  
124    Colombo            127  
125    Colombo            128  

[51109 rows x 8 columns]

In [47]:
output['over'] = output['ball'].apply(lambda x:str(x).split(".")[0])
output['ball_no'] = output['ball'].apply(lambda x:str(x).split(".")[1])
output

C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/2944276496.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['over'] = output['ball'].apply(lambda x:str(x).split(".")[0])
C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/2944276496.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['ball_no'] = output['ball'].apply(lambda x:str(x).split(".")[1])


match_id batting_team bowling_team  ball  runs player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                0   
1           1    Australia    Sri Lanka   0.2     0                0   
2           1    Australia    Sri Lanka   0.3     1                0   
3           1    Australia    Sri Lanka   0.4     2                0   
4           1    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       963    Sri Lanka    Australia  19.3     1                0   
122       963    Sri Lanka    Australia  19.4     0                0   
123       963    Sri Lanka    Australia  19.5     0      DM de Silva   
124       963    Sri Lanka    Australia  19.6     2                0   
125       963    Sri Lanka    Australia  19.7     1                0   

          city  current_score over ball_no  
0    Melbourne              0    0       1  
1    Melbourne              0    0       2  
2    Melbourne              1    0       3  
3    Melbourne              3    0       4  
4    Melbourne              3    0       5  
..         ...            ...  ...     ...  
121    Colombo            125   19       3  
122    Colombo            125   19       4  
123    Colombo            125   19       5  
124    Colombo            127   19       6  
125    Colombo            128   19       7  

[51109 rows x 10 columns]

In [48]:
output['balls_bowled'] = (output['over'].astype('int')*6) + output['ball_no'].astype('int')
output

C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/4047150941.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['balls_bowled'] = (output['over'].astype('int')*6) + output['ball_no'].astype('int')


match_id batting_team bowling_team  ball  runs player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                0   
1           1    Australia    Sri Lanka   0.2     0                0   
2           1    Australia    Sri Lanka   0.3     1                0   
3           1    Australia    Sri Lanka   0.4     2                0   
4           1    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       963    Sri Lanka    Australia  19.3     1                0   
122       963    Sri Lanka    Australia  19.4     0                0   
123       963    Sri Lanka    Australia  19.5     0      DM de Silva   
124       963    Sri Lanka    Australia  19.6     2                0   
125       963    Sri Lanka    Australia  19.7     1                0   

          city  current_score over ball_no  balls_bowled  
0    Melbourne              0    0       1             1  
1    Melbourne              0    0       2             2  
2    Melbourne              1    0       3             3  
3    Melbourne              3    0       4             4  
4    Melbourne              3    0       5             5  
..         ...            ...  ...     ...           ...  
121    Colombo            125   19       3           117  
122    Colombo            125   19       4           118  
123    Colombo            125   19       5           119  
124    Colombo            127   19       6           120  
125    Colombo            128   19       7           121  

[51109 rows x 11 columns]

In [49]:
output['balls_left'] = 120 - output['balls_bowled']
output['balls_left'] = output['balls_left'].apply(lambda x:0 if x<0 else x)
output

C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/3456931282.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['balls_left'] = 120 - output['balls_bowled']
C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/3456931282.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['balls_left'] = output['balls_left'].apply(lambda x:0 if x<0 else x)


match_id batting_team bowling_team  ball  runs player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                0   
1           1    Australia    Sri Lanka   0.2     0                0   
2           1    Australia    Sri Lanka   0.3     1                0   
3           1    Australia    Sri Lanka   0.4     2                0   
4           1    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       963    Sri Lanka    Australia  19.3     1                0   
122       963    Sri Lanka    Australia  19.4     0                0   
123       963    Sri Lanka    Australia  19.5     0      DM de Silva   
124       963    Sri Lanka    Australia  19.6     2                0   
125       963    Sri Lanka    Australia  19.7     1                0   

          city  current_score over ball_no  balls_bowled  balls_left  
0    Melbourne              0    0       1             1         119  
1    Melbourne              0    0       2             2         118  
2    Melbourne              1    0       3             3         117  
3    Melbourne              3    0       4             4         116  
4    Melbourne              3    0       5             5         115  
..         ...            ...  ...     ...           ...         ...  
121    Colombo            125   19       3           117           3  
122    Colombo            125   19       4           118           2  
123    Colombo            125   19       5           119           1  
124    Colombo            127   19       6           120           0  
125    Colombo            128   19       7           121           0  

[51109 rows x 12 columns]

As we want to know how many wickets are left after a certain period or overs. For that we need to extract this data from above dataset as in the player_dismiised column it is given as we can see that if a wicket down than that **players name is given or else the value is 0**. So with it we will extract wickets left data with the following code

In [50]:
output['player_dismissed'] = output['player_dismissed'].apply(lambda x:0 if x=='0' else 1)
output['player_dismissed'] = output['player_dismissed'].astype('int')
output['player_dismissed'] = output.groupby('match_id').cumsum()['player_dismissed']
output['wickets_left'] = 10 - output['player_dismissed']

C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/2957949265.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['player_dismissed'] = output['player_dismissed'].apply(lambda x:0 if x=='0' else 1)
C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/2957949265.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['player_dismissed'] = output['player_dismissed'].astype('int')
C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/2957949265.py:3: SettingWithCopyWarning: 
A value is trying to be set on

In [51]:
output

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                 0   
1           1    Australia    Sri Lanka   0.2     0                 0   
2           1    Australia    Sri Lanka   0.3     1                 0   
3           1    Australia    Sri Lanka   0.4     2                 0   
4           1    Australia    Sri Lanka   0.5     0                 0   
..        ...          ...          ...   ...   ...               ...   
121       963    Sri Lanka    Australia  19.3     1                 8   
122       963    Sri Lanka    Australia  19.4     0                 8   
123       963    Sri Lanka    Australia  19.5     0                 9   
124       963    Sri Lanka    Australia  19.6     2                 9   
125       963    Sri Lanka    Australia  19.7     1                 9   

          city  current_score over ball_no  balls_bowled  balls_left  \
0    Melbourne              0    0       1             1         119   
1    Melbourne              0    0       2             2         118   
2    Melbourne              1    0       3             3         117   
3    Melbourne              3    0       4             4         116   
4    Melbourne              3    0       5             5         115   
..         ...            ...  ...     ...           ...         ...   
121    Colombo            125   19       3           117           3   
122    Colombo            125   19       4           118           2   
123    Colombo            125   19       5           119           1   
124    Colombo            127   19       6           120           0   
125    Colombo            128   19       7           121           0   

     wickets_left  
0              10  
1              10  
2              10  
3              10  
4              10  
..            ...  
121             2  
122             2  
123             1  
124             1  
125             1  

[51109 rows x 13 columns]

We want Current run rate also as one of our feauture so we will extract that also and add that current run rate **(crr)** column into our dataset

In [54]:
output['crr'] = (output['current_score']*6)/output['balls_bowled']

C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/764537999.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['crr'] = (output['current_score']*6)/output['balls_bowled']


In [55]:
output

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                 0   
1           1    Australia    Sri Lanka   0.2     0                 0   
2           1    Australia    Sri Lanka   0.3     1                 0   
3           1    Australia    Sri Lanka   0.4     2                 0   
4           1    Australia    Sri Lanka   0.5     0                 0   
..        ...          ...          ...   ...   ...               ...   
121       963    Sri Lanka    Australia  19.3     1                 8   
122       963    Sri Lanka    Australia  19.4     0                 8   
123       963    Sri Lanka    Australia  19.5     0                 9   
124       963    Sri Lanka    Australia  19.6     2                 9   
125       963    Sri Lanka    Australia  19.7     1                 9   

          city  current_score over ball_no  balls_bowled  balls_left  \
0    Melbourne              0    0       1             1         119   
1    Melbourne              0    0       2             2         118   
2    Melbourne              1    0       3             3         117   
3    Melbourne              3    0       4             4         116   
4    Melbourne              3    0       5             5         115   
..         ...            ...  ...     ...           ...         ...   
121    Colombo            125   19       3           117           3   
122    Colombo            125   19       4           118           2   
123    Colombo            125   19       5           119           1   
124    Colombo            127   19       6           120           0   
125    Colombo            128   19       7           121           0   

     wickets_left       crr  
0              10  0.000000  
1              10  0.000000  
2              10  2.000000  
3              10  4.500000  
4              10  3.600000  
..            ...       ...  
121             2  6.410256  
122             2  6.355932  
123             1  6.302521  
124             1  6.350000  
125             1  6.347107  

[51109 rows x 14 columns]

**Cricket is a game of momentum** and to predict something in cricket specially score, it is very important to know where the momentum is going in terms of runs, wickets or even in fielding. Here we are using last three overs runs to know the momentum. We can also use 2 overs or even 5 overs. We will do that with the help of following code.

In [56]:
groups = output.groupby('match_id')

match_ids = output['match_id'].unique()
last_three = []
for id in match_ids:
    last_three.extend(groups.get_group(id).rolling(window=18).sum()['runs'].values.tolist())  
    # 18 because 3 overs means 18 balls 

In [57]:
output['last_three'] = last_three
output

C:\Users\Mahesh\AppData\Local\Temp/ipykernel_12460/1360788657.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['last_three'] = last_three


match_id batting_team bowling_team  ball  runs  player_dismissed  \
0           1    Australia    Sri Lanka   0.1     0                 0   
1           1    Australia    Sri Lanka   0.2     0                 0   
2           1    Australia    Sri Lanka   0.3     1                 0   
3           1    Australia    Sri Lanka   0.4     2                 0   
4           1    Australia    Sri Lanka   0.5     0                 0   
..        ...          ...          ...   ...   ...               ...   
121       963    Sri Lanka    Australia  19.3     1                 8   
122       963    Sri Lanka    Australia  19.4     0                 8   
123       963    Sri Lanka    Australia  19.5     0                 9   
124       963    Sri Lanka    Australia  19.6     2                 9   
125       963    Sri Lanka    Australia  19.7     1                 9   

          city  current_score over ball_no  balls_bowled  balls_left  \
0    Melbourne              0    0       1             1         119   
1    Melbourne              0    0       2             2         118   
2    Melbourne              1    0       3             3         117   
3    Melbourne              3    0       4             4         116   
4    Melbourne              3    0       5             5         115   
..         ...            ...  ...     ...           ...         ...   
121    Colombo            125   19       3           117           3   
122    Colombo            125   19       4           118           2   
123    Colombo            125   19       5           119           1   
124    Colombo            127   19       6           120           0   
125    Colombo            128   19       7           121           0   

     wickets_left       crr  last_three  
0              10  0.000000         NaN  
1              10  0.000000         NaN  
2              10  2.000000         NaN  
3              10  4.500000         NaN  
4              10  3.600000         NaN  
..            ...       ...         ...  
121             2  6.410256        20.0  
122             2  6.355932        19.0  
123             1  6.302521        18.0  
124             1  6.350000        16.0  
125             1  6.347107        16.0  

[51109 rows x 15 columns]

Now what we will see is that first 3 overs or 18 balls of all mathces will have NaN value because we are adding last three overs data so that will be available after 3 overs only. **Now again from the above data we will extract all the relevant columns we want in our model**.Our target is to get total runs at the end of the innings so that should also be there in our dataset.

In [58]:
#code for total runs
final_df = output.groupby('match_id').sum()['runs'].reset_index().merge(output,on='match_id')

In [59]:
final_df

match_id  runs_x batting_team bowling_team  ball  runs_y  \
0             1     168    Australia    Sri Lanka   0.1       0   
1             1     168    Australia    Sri Lanka   0.2       0   
2             1     168    Australia    Sri Lanka   0.3       1   
3             1     168    Australia    Sri Lanka   0.4       2   
4             1     168    Australia    Sri Lanka   0.5       0   
...         ...     ...          ...          ...   ...     ...   
51104       963     128    Sri Lanka    Australia  19.3       1   
51105       963     128    Sri Lanka    Australia  19.4       0   
51106       963     128    Sri Lanka    Australia  19.5       0   
51107       963     128    Sri Lanka    Australia  19.6       2   
51108       963     128    Sri Lanka    Australia  19.7       1   

       player_dismissed       city  current_score over ball_no  balls_bowled  \
0                     0  Melbourne              0    0       1             1   
1                     0  Melbourne              0    0       2             2   
2                     0  Melbourne              1    0       3             3   
3                     0  Melbourne              3    0       4             4   
4                     0  Melbourne              3    0       5             5   
...                 ...        ...            ...  ...     ...           ...   
51104                 8    Colombo            125   19       3           117   
51105                 8    Colombo            125   19       4           118   
51106                 9    Colombo            125   19       5           119   
51107                 9    Colombo            127   19       6           120   
51108                 9    Colombo            128   19       7           121   

       balls_left  wickets_left       crr  last_three  
0             119            10  0.000000         NaN  
1             118            10  0.000000         NaN  
2             117            10  2.000000         NaN  
3             116            10  4.500000         NaN  
4             115            10  3.600000         NaN  
...           ...           ...       ...         ...  
51104           3             2  6.410256        20.0  
51105           2             2  6.355932        19.0  
51106           1             1  6.302521        18.0  
51107           0             1  6.350000        16.0  
51108           0             1  6.347107        16.0  

[51109 rows x 16 columns]

runs_x in the above dataset is showing total runs at the end of innings and runs_y is showing run in that particular bowl

In [60]:
#All the columns or data that i am using for my model
final_df=final_df[['batting_team','bowling_team','city','current_score','balls_left','wickets_left','crr','last_three','runs_x']]

In [61]:
final_df.isnull().sum()

batting_team        0
bowling_team        0
city                0
current_score       0
balls_left          0
wickets_left        0
crr                 0
last_three       7141
runs_x              0
dtype: int64

In [62]:
#remove rows with missing data
final_df.dropna(inplace=True)

C:\Users\Mahesh\anaconda3\lib\site-packages\pandas\util\_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [63]:
final_df.isnull().sum()

batting_team     0
bowling_team     0
city             0
current_score    0
balls_left       0
wickets_left     0
crr              0
last_three       0
runs_x           0
dtype: int64

We should try from our end that there is no bias developed due to our data extraction and analysis work. So for that we will **Suffle the data**.

In [64]:
final_df = final_df.sample(final_df.shape[0])

In [65]:
final_df

batting_team bowling_team          city  current_score  balls_left  \
41021      Pakistan   Bangladesh        Mirpur            135          28   
42600  South Africa  West Indies  Johannesburg            113          62   
23153  South Africa      England     Centurion             72          81   
29682  South Africa    Australia     Cape Town             58          62   
18366         India     Pakistan  Johannesburg            111          29   
...             ...          ...           ...            ...         ...   
37380         India     Pakistan     Bangalore            103          36   
48413    Bangladesh    Australia     Bangalore            104          30   
24261      Pakistan   Bangladesh      St Lucia             67          70   
27498      Pakistan  New Zealand         Dubai            114          31   
37976   West Indies      England      Barbados             76          62   

       wickets_left        crr  last_three  runs_x  
41021             7   8.804348        36.0     190  
42600             8  11.689655        42.0     231  
23153            10  11.076923        45.0     241  
29682             8   6.000000        19.0     146  
18366             7   7.318681        21.0     157  
...             ...        ...         ...     ...  
37380             8   7.357143        25.0     133  
48413             6   6.933333        24.0     156  
24261            10   8.040000        25.0     172  
27498             6   7.685393        24.0     153  
37976             9   7.862069        14.0     170  

[43968 rows x 9 columns]

Finally we are at the door of model building as we have done data extraction and feauture extraction as well. Now further tasks are to split data into training and testing dataset and build model and do the prediction.

In [66]:
X = final_df.drop(columns=['runs_x'])
y = final_df['runs_x']
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=1)

In [67]:
X_train

batting_team  bowling_team          city  current_score  balls_left  \
18290        India      Pakistan  Johannesburg             25         103   
41600    Sri Lanka   West Indies        Mirpur             53          74   
49489  Afghanistan   West Indies        Nagpur             50          62   
12163    Australia       England   Southampton             93          36   
23621  New Zealand     Sri Lanka       Colombo             19         103   
...            ...           ...           ...            ...         ...   
8231   New Zealand         India      Auckland             39          92   
45038    Australia  South Africa        Durban            147          12   
6211     Sri Lanka    Bangladesh       Colombo            127          19   
45656    Australia         India        Sydney            169          16   
9124   New Zealand      Pakistan         Dubai             62          73   

       wickets_left       crr  last_three  
18290             9  8.823529        24.0  
41600             7  6.913043        11.0  
49489             7  5.172414        14.0  
12163             5  6.642857        24.0  
23621             9  6.705882        19.0  
...             ...       ...         ...  
8231              9  8.357143        35.0  
45038             3  8.166667        34.0  
6211              5  7.544554        38.0  
45656             6  9.750000        34.0  
9124              8  7.914894        16.0  

[35174 rows x 8 columns]

In [69]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.metrics import r2_score,mean_absolute_error

In [70]:
# Because we have three columns in our dataset are string so we need column transformer
trf = ColumnTransformer([
    ('trf',OneHotEncoder(sparse=False,drop='first'),['batting_team','bowling_team','city'])
]
,remainder='passthrough')

In [71]:
# Now we are creating a sklearn pipeline where first data will transform as per trf then it will we scaled and finally xgboost
pipe = Pipeline(steps=[
    ('step1',trf),
    ('step2',StandardScaler()),
    ('step3',XGBRegressor(n_estimators=1000,learning_rate=0.2,max_depth=12,random_state=1))
])


In [76]:
#training the model and seeing mae and r^2 score
pipe.fit(X_train,y_train)
y_pred = pipe.predict(X_test)
print(r2_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))

0.9855010466360374
1.8472508887472494


So it is giving good r^2 value and we can save it so will save the model so that we can deploy it.

In [77]:
pickle.dump(pipe,open('pipe.pkl','wb'))

In [78]:
import xgboost
xgboost.__version__

'1.5.0'